### PR TITLE
Feat: access questions by the URL

### PR DESCRIPTION
**Pull request by EmileRolley**
Implements #1020.

## Specification

Actually, there already is a system to manage URL allowing to use sub-simulators such as: https://nosgestesclimat.fr/simulateur/transport/empreinte. We decide to keep this possibility to be able to use a sub-simulator for each rules. We had the possibility to navigate through questions via the `question` search param.

**It's only possible to be redirected to a _valid_ rule** (i.e. it is included in the loaded model rules, contains a `question` and isn't a _mosaic child_). If the URL doesn't correspond to a _valid_ rule, a redirection is done to the first parent _valid_ rule (and if such a rule doesn't exist, to the root `/simulateur/bilan`).

### Examples

* The URL corresponding to the question `logement . saisie habitants` of the main simulator is `/simulateur/bilan?question=logement.saisie-habitants`
* The URL corresponding to the question  `logement . électricité . consommation` of the sub-simulator `logement` is `/simulateur/logement?question=logement.électricité.consommation`. Indeed, we only want to access to the sub rules corresponding to the category of the sub-simulator.

## Questions

1. What should be the wanted behavior when arriving with a custom question URL with the tutorial not skipped yet?
    > Currently, it will redirect to `/tutoriel` then to `/simulateur/bilan`.

   **We decided with @Clemog and @bjlaa to not display the tutorial when visiting a non-root url.**

2. What should be the wanted behavior for mosaic questions?
    > Currently, you can access to a mosaic question from the base rule and all of the rules composing the mosaic question (as `C-k`). 

    **We decided with @Clemog and @bjlaa to keep an unique URL for each mosaic question.**

3. What should be the wanted behavior for the navigator `go back` button? 

### Core logic
 
* Use `navigate()` instead of `dispatch(goToQuestion())` in `Conversation.tsx`. The logic is handled in the `Simulateur.tsx` file (previously `Simulateur.js`).
* Use a new attribute `fromRule` in the `TutorialState` of the action `skipTutorial` in order to track redirection to `/tutorial` from a specific rule URL: see [`rootReducer.ts`](https://github.com/datagir/nosgestesclimat-site/pull/1037/files#diff-a39cffebe7cedc5c15cf1bfac946d399b9b0f70118a5f00ea86034c52035dcbc), [`actions.ts`](https://github.com/datagir/nosgestesclimat-site/pull/1037/files#diff-e62eba865dfb27710138be0b324a9bba1d10c33f2ef8f4dea4abfd2ffa124d9a), [`Tutorial.tsx`](https://github.com/datagir/nosgestesclimat-site/pull/1037/files#diff-a349258badcef80e960da7c1e60f3bb371c1a07e4442c871787090884e46bd5f)
* Props type refactoring for [`Simulation.tsx`](https://github.com/datagir/nosgestesclimat-site/pull/1037/files#diff-ed679b292892693c542ddc82b996e92ad748984d7e9e321e3509be207e16e189),  [`Conversation.tsx`](https://github.com/datagir/nosgestesclimat-site/pull/1037/files#diff-37589bb8a6d3a1d25ee4c229e646eb5e2c5ee7e3e90c6fdd8398bff75c28220fR67-R73)
  * Add new field [`isFromActionCard`](https://github.com/datagir/nosgestesclimat-site/pull/1037/files#diff-37589bb8a6d3a1d25ee4c229e646eb5e2c5ee7e3e90c6fdd8398bff75c28220fR80) to use `dispatch(goToQuestion)` instead of the new `navigate` logic in [`ActionConversation.tsx`](https://github.com/datagir/nosgestesclimat-site/pull/1037/files#diff-4408e691b6b39e24377fe9b0d78e6e0f024bef914d7a716f78615964c1696924). The logic is therefore factorized in [`goToQuestionOrNavigate`](https://github.com/datagir/nosgestesclimat-site/pull/1037/files#diff-6f03b112e729af79352ef520cd08d06ba2b7c7ca9df4cbc39b3618f01de53f63R64-R79)

### Tests

* New e2e test file: `url-redirection.cy.js`
* New e2e test in `question-order.cy.js`
* Code factorization of the `test-completion` tests.

### Refactoring and cleaning

* Refactoring (i.e. add or complete existing type defs, factorize code, use `@/...` import alias, change `.js` file to `.ts`) for modified files.
* In particular, try to extract util functions in [`conversationUtils.ts`](https://github.com/datagir/nosgestesclimat-site/pull/1037/files#diff-6f03b112e729af79352ef520cd08d06ba2b7c7ca9df4cbc39b3618f01de53f63)
* New [`MosaicInfos`](https://github.com/datagir/nosgestesclimat-site/pull/1037/files#diff-5dda4a4cab6b6da7d71cd49165801936b98483fdc3cf790a0a5a06efe00292f1R66-R70) type.
* Change `publicodesUtils.js` to [`publicodesUtils.ts`](https://github.com/datagir/nosgestesclimat-site/pull/1037/files#diff-b5a534af6b9823b1b176630d86d3ba32f0fb814d651ace9cc0927b9bfda26d50) and introduce new type.
* Start to migrate scripts to TypeScript by introducing `ts-node` in dev dependencies. 

### Scripts

* Add `simulateur/bilan` question rule URLs to the sitemap -- #1019

### What's next to do (in an other PR)

* [ ] Ctrl-K should use navigate to + (filter mosaic child questions?)
* [x] Support category sub-simulator

### CODE REVIEW

**Comment by bjlaa** [comment]
@jeremiecook et Emile, est-ce que le bouton précédent dans le questionnaire fait toujours aussi sens qu'avant la navigation via URL ?

File: source/components/publicodesUtils.tsx [hunk 1]
```diff
@@ -24,28 +84,37 @@ export const FullName = ({ dottedName }) => (
 export const title = (rule) =>
 	rule.titre ||
 	capitalise0(splitName(rule.dottedName)[splitName(rule.dottedName).length - 1])
+
 // Publicodes's % unit is strangely handlded
 // the nodeValue is * 100 to account for the unit
 // hence we divide it by 100 and drop the unit
-export const correctValue = (evaluated) => {
+export function correctValue(evaluated: EvaluatedNode): number | undefined {
```
Comment (bjlaa) [inline source/components/publicodesUtils.tsx:91]: Pas fan des fonctions fléchées à ce que je vois ? 😄

File: source/components/publicodesUtils.tsx [hunk 2]
```diff
@@ -24,28 +84,37 @@ export const FullName = ({ dottedName }) => (
...
+	if (nodeValue == undefined || typeof nodeValue !== 'number') {
+		return undefined
+	}
+
 	const result = unit?.numerators.includes('%') ? nodeValue / 100 : nodeValue
 	return result
 }
 
-const ruleSumNode = (rules, rule) => {
+function ruleSumNode(rules: any, rule: RuleNode): string[] | undefined {
```
Comment (bjlaa) [inline source/components/publicodesUtils.tsx:102]: Est-ce qu'on ne pourrait pas passer le type `RuleNodes` au lieu de `any` ici ?

File: source/components/publicodesUtils.tsx [hunk 3]
```diff
@@ -24,28 +84,37 @@ export const FullName = ({ dottedName }) => (
...
+	return formula['somme'].map((name: string) =>
 		coreUtils.disambiguateReference(rules, rule.dottedName, name)
 	)
 }
 
 export const extractCategoriesNamespaces = (
-	rules,
-	engine,
-	parentRule = 'bilan'
+	rules: { [x: string]: { icônes: any; couleur: any } },
```
Comment (bjlaa) [inline source/components/publicodesUtils.tsx:115]: c'est un point de détail mais j'aime bien passer `key` à la place de `x` je trouve ça plus explicite

File: source/components/utils/Meta.tsx [hunk 4]
```diff
@@ -22,7 +23,7 @@ export default function Meta({
 	const { pathname } = useLocation()
 	return (
 		<Helmet>
-			<title>{title} - Nos Gestes Climat</title>
+			<title>{`${title} - Nos Gestes Climat`}</title>
```
Comment (bjlaa) [inline source/components/utils/Meta.tsx:26]: warum ?

Comment (bjlaa) [reply source/components/utils/Meta.tsx:26]: blague à part, ça change pas le résultat du coup curieux de connaitre tes motivations à faire ce changement

File: source/sites/publicodes/BandeauContribuer.js [hunk 5]
```diff
@@ -18,33 +18,34 @@ export default () => {
 		show && setTimeout(() => setVisible(true), 8000)
 	}, [show])
 
+	if (!show || !visible) {
```
Comment (bjlaa) [inline source/sites/publicodes/BandeauContribuer.js:21]: 👍

File: source/sites/publicodes/pages/QuestionList.tsx [hunk 6]
```diff
@@ -85,12 +89,8 @@ export default () => {
...
 
-const getQuestionType = (engine, rules, rule) => {
-	const ruleMosaicInfos = getRelatedMosaicInfosIfExists(
-		engine,
-		rules,
-		rule.dottedName
-	)
+const getQuestionType = (rules: RulesNodes, rule: NGCRuleNode) => {
+	const ruleMosaicInfos = getRelatedMosaicInfosIfExists(rules, rule.dottedName)
 	console.log(ruleMosaicInfos)
```
Comment (bjlaa) [inline source/sites/publicodes/pages/QuestionList.tsx:94]: oubli ?

**Comment by EmileRolley** [comment]
> @jeremiecook et Emile, est-ce que le bouton précédent dans le questionnaire fait toujours aussi sens qu'avant la navigation via URL ?

Je en vois pas pourquoi ça le remettrait en question ? Ca ne devrait pas changer beaucoup de chose à l'expérience utilsateur.ice non ? C'est plus une amélioration côté stats et SEO.

File: source/sites/publicodes/pages/DocumentationPage.tsx [hunk 7]
```diff
@@ -20,8 +21,8 @@ export default ({ engine: givenEngine }) => {
 	return (
 		<DocumentationStyle>
 			<RulePage
-				language={i18n.language}
-				rulePath={url}
+				language={i18n.language as 'fr' | 'en'}
```
[the remark this answers is not in the document]
Comment (EmileRolley) [reply source/sites/publicodes/pages/DocumentationPage.tsx:24]: Non pas vraiment, on utilise l'enum de `Lang` que j'avais faite au tout début du travail sur la traduction. Cependant je n'ai pas l'impression que ce soit dans les pratiques standard de TS, ça vaudrait peut-être le coup de passer sur un union type like avec `'fr' | 'en'` à la place.

**Comment by bjlaa** [comment]
> Je en vois pas pourquoi ça le remettrait en question ? Ca ne devrait pas changer beaucoup de chose à l'expérience utilsateur.ice non ? C'est plus une amélioration côté stats et SEO.

Oui tu as raison, en plus côté mobile ça fait plus app ! J'avais pas encore vu que tu avais implémenté le `navigate` sur le bouton Précédent, c'est super cool !

File: source/components/utils/Meta.tsx (hunk 4 above)
Comment (laem) [reply source/components/utils/Meta.tsx:26]: Globalement, je crois qu'il vaut mieux utiliser Meta.tsx

**Comment by laem** [comment]
J'ai fait une revue sans regarder le code. Quelques éléments :


1) Pour moi le fait de faire disparaitre la possibilité de lancer des simulateurs sur chaque variable publicodes est bloquante. Par exemple https://deploy-preview-1037--nosgestesclimat.netlify.app/simulateur/transport/train. 

2) Ça rejoint le point 1), je trouve qu'il y a confusion entre la variable de calcul qui déclenche le simulateur ("bilan" souvent, mais potentiellement n'importe quelle variable), et l'état du questionnaire, qui s'enchaînent à la suite dans l'URL avec des `/`. https://deploy-preview-1037--nosgestesclimat.netlify.app/simulateur/bilan/transport/voiture/km. 

Alors que ce sont deux choses différentes. La première est la page qu'on veut visiter, et la deuxième est un état de la page, ou un filtre. Du coup, l'usage des query params me semble adaptée.

Pourquoi pas /simulateur/bilan?question=transport/voiture/km ?
Ou encore /simulateur/transport/empreinte?question=transport/deux-roues-thermique/usager. 

Les query params sont bien pris en compte comme des URL de pages distinctes par les moteurs de recherche (contrairement aux #), et me semblent accepter les /, d'après [la RFC](https://stackoverflow.com/questions/68881593/do-i-need-to-escape-forward-slashes-in-a-url-query-string) et mes tests sur chromium et firefox. 

3) je ne suis pas convaincu par le piège de boutons précédent. Après avoir répondu à 10 questions, je devrais cliquer 10 fois sur précédent pour revenir à la page d'accueil, si je n'ai pas envie d'utiliser le menu. Et idem sans raccourci si je partais de gouvernement.fr et que je voulais revenir sur gouvernement.fr. Ça ne me semble pas être l'objectif du bouton back. Comme expliqué ici https://github.com/datagir/nosgestesclimat-site/issues/463. Mais c'est pas bloquant, on peut aussi tester comme ça, mais on ne captera pas les frustrations potentielles des utilisateurs. 

Ce point rejoint le 2) : les questions du simulateur ne sont pas des pages (au sens page Web, par exemple un article de blog), mais un état d'une page, le simulateur choisi.

File: source/components/conversation/conversationUtils.ts [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+	// to find the first question that is not in the current mosaic
+	return previousAnswers[currentQuestionIndex - 1]
+}
+
+export function goToQuestionOrNavigate(
+	question: DottedName,
+	focusedCategory: string | null,
+	toUse: { dispatch: Dispatch<AnyAction> } | { navigate: NavigateFunction }
+): void {
+	if (toUse['navigate'] != undefined) {
```
Comment (bjlaa) [inline source/components/conversation/conversationUtils.ts:69]: Pourquoi choisir de passer un objet pour `toUse` et pas directement une fonction ? Je trouve que ça prête un poil à confusion les appels à toUse comme `toUse['navigate']` 😕

File: source/components/conversation/conversationUtils.ts [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+export function focusByCategory(
+	questions: DottedName[],
+	focusedCategory: string | null
+) {
+	if (!focusedCategory) {
+		return questions
+	}
+	const filtered = questions.filter((q) => q.indexOf(focusedCategory) === 0)
```
Comment (bjlaa) [inline source/components/conversation/conversationUtils.ts:89]: Pas sûr de bien comprendre ici : si questions est un array de `dottednames`, que fait `indexOf` ici ?

File: source/components/publicodesUtils.tsx [hunk 10]
```diff
@@ -24,31 +88,43 @@ export const FullName = ({ dottedName }) => (
...
+	return formula['somme'].map((name: string) =>
 		coreUtils.disambiguateReference(rules, rule.dottedName, name)
 	)
 }
 
 export const extractCategoriesNamespaces = (
-	rules,
-	engine,
-	parentRule = 'bilan'
+	rules: { [key: string]: { icônes: any; couleur: any } },
```
Comment (bjlaa) [inline source/components/publicodesUtils.tsx:122]: accent circonflexe dans une variable 🤢

File: source/components/publicodesUtils.tsx [hunk 11]
```diff
@@ -105,10 +181,10 @@ export const extractCategories = (
 	}
 
 	const categories = sumNodes.map((dottedName) => {
-		const node = engine.evaluate(dottedName)
+		const node = engine.evaluate(dottedName) as Category
 		const { icônes, couleur, abréviation } = rules[dottedName]
 		const split = splitName(dottedName),
-			parent = split.length > 1 && split[0]
+			parent = split.length > 1 ? split[0] : ''
```
Comment (bjlaa) [inline source/components/publicodesUtils.tsx:187]: Par curiosité que change cette modif ? L'app pétait dans certains cas ?

File: source/components/publicodesUtils.tsx [hunk 12]
```diff
@@ -148,7 +229,10 @@ export const getSubcategories = (rules, category, engine, sort) => {
 			? subCategories.map((el) => ({
```
Comment (bjlaa) [inline source/components/publicodesUtils.tsx:229]: Ça me parait un peu bizarre d'avoir un cas particulier dans un méthode au nom si général 🤔 pas de ton fait je sais bien !

File: source/sites/publicodes/Simulateur.tsx [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+	'alimentation',
+	'logement',
+	'services-sociétaux',
+	'divers',
+]
+
+const Simulateur = () => {
+	const dispatch = useDispatch()
+	const urlParams = useParams()
+	const path = urlParams['*']?.split('/')
```
Comment (bjlaa) [inline source/sites/publicodes/Simulateur.tsx:72]:
```suggestion
	const pathElements = urlParams['*']?.split('/')
```
Pour laisser comprendre que c'est un array avec plusieurs éléments ?

File: source/sites/publicodes/Simulateur.tsx [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+	if (!path) {
+		return <Navigate to={`/simulateur/${MODEL_ROOT_RULE_NAME}`} replace />
+	}
+
+	const category = path[0]
+	const isMainSimulation = isRootRule(category)
```
Comment (bjlaa) [inline source/sites/publicodes/Simulateur.tsx:78]: Désolé je vais un peu relou mais je comprends pas bien ce qu'est cette constante ; je me dis que tant qu'on a les mains dans le camboui autant rendre ce code le plus compréhensible possible !

File: source/sites/publicodes/Simulateur.tsx [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+	const selectedRuleNameURLPath = path.slice(1)
+
+	if (!availableCategories.includes(category)) {
+		console.log(
```
Comment (bjlaa) [inline source/sites/publicodes/Simulateur.tsx:82]: Ici est-ce qu'il ne vaudrait pas mieux throw une erreur qui pourrait être récupérée par Sentry, au cas où ?

File: source/sites/publicodes/chart/InlineCategoryChart.tsx [hunk 16]
```diff
@@ -24,26 +26,25 @@ const SpecializedVisualisation = React.lazy(
 		)
 )
 
+function mapAbreviation(rules: any, engine: Engine) {
+	return extractCategories(rules, engine).map((category) => {
```
Comment (bjlaa) [inline source/sites/publicodes/chart/InlineCategoryChart.tsx:30]: Je sais pas ce que tu en penses mais dans ce cas je me dis qu'ajouter une constante permettrait de gagner en clarté :
```
const categories = extractCategories(rules, engine)
return categories.map((category) => {...
```

File: source/sites/publicodes/chart/InlineCategoryChart.tsx [hunk 17]
```diff
@@ -24,26 +26,25 @@ const SpecializedVisualisation = React.lazy(
...
+	const tutorials = useSelector((state: AppState) => state.tutorials)
 
 	useEffect(() => {
-		setCategories(
-			extractCategories(rules, engine).map((category) => ({
-				...category,
-				abbreviation: rules[category.dottedName].abréviation,
-			}))
-		)
+		setCategories(mapAbreviation(rules, engine))
```
Comment (bjlaa) [inline source/sites/publicodes/chart/InlineCategoryChart.tsx:47]: Pourquoi faut-il remaper les abbréviations à chaque modification de la situation ?

Comment (bjlaa) [reply source/sites/publicodes/chart/InlineCategoryChart.tsx:47]: Pas moyen d'utiliser un useMemo ou d'enregistrer le résultat dans une ref ?

File: source/sites/publicodes/pages/DocumentationPage.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (bjlaa) [reply source/sites/publicodes/pages/DocumentationPage.tsx:24]: Oui un type serait de bon aloi 😄

File: source/sites/publicodes/Simulateur.tsx [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+			// Case where we previously visited a specific rule URL and we come back
+			// (the tutorial was skipped) to the simulator root URL,
+			// we want to display the tutorial
+			(!isSpecificRule(selectedRuleDottedName) && tutorials.fromRule == 'skip'))
+
+	return (
+		<div>
+			<Meta
+				title={evaluation.rawNode?.title}
+				description={evaluation.rawNode?.description}
```
Comment (bjlaa) [inline source/sites/publicodes/Simulateur.tsx:140]: On sait quelle taille fait la description ? Normalement il faut pas dépasser les 160 caractères

File: source/sites/publicodes/Simulateur.tsx [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+	categoryName: string,
+	rules: NGCRulesNodes
+): SelectedRuleInfos {
+	const navigate = useNavigate()
+	const isValidRule = (ruleName: DottedName) => {
+		if (rules == undefined) {
+			return false
+		}
+		const rule = rules[ruleName]
+		const isAQuestion =
```
Comment (bjlaa) [inline source/sites/publicodes/Simulateur.tsx:218]: Pas obligé de changer ici mais d'expérience le "a" saute souvent dans les noms de constantes boolean type `isMachin` 😄

File: source/sites/publicodes/Simulateur.tsx [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+			rule != undefined && 'rawNode' in rule && 'question' in rule.rawNode
+		const isAMosaicChild = rule != undefined && isMosaicChild(rules, ruleName)
+
+		return isAQuestion && !isAMosaicChild
+	}
+
+	if (selectedRuleName != '' && !isValidRule(selectedRuleName)) {
```
Comment (bjlaa) [inline source/sites/publicodes/Simulateur.tsx:225]: Exemple ici d'ailleurs !

File: source/sites/publicodes/Simulateur.tsx [hunk 21]
```diff
@@ -0,0 +1,339 @@
...
+		while (selectedRuleName != '' && !isValidRule(selectedRuleName)) {
+			const parentRuleName = utils.ruleParent(selectedRuleName)
+			console.log(
+				`Unknown question ${selectedRuleName}, trying ${parentRuleName}...`
+			)
+			selectedRuleName = parentRuleName
+		}
+
+		if (selectedRuleName == '') {
+			console.log(
```
Comment (bjlaa) [inline source/sites/publicodes/Simulateur.tsx:235]: Même remarque, un throw pour nourrir notre Sentry ?

**Comment by bjlaa** [comment]
> Pour moi le fait de faire disparaitre la possibilité de lancer des simulateurs sur chaque variable publicodes est bloquante. Par exemple https://deploy-preview-1037--nosgestesclimat.netlify.app/simulateur/transport/train.

Ah oui c'est voulu ça ? Il me semblait que l'on s'était dit qu'on voulait justement les garder ces simulateurs par catégorie 🤔 

> Ça rejoint le point 1), je trouve qu'il y a confusion entre la variable de calcul qui déclenche le simulateur ("bilan" souvent, mais potentiellement n'importe quelle variable), et l'état du questionnaire, qui s'enchaînent à la suite dans l'URL avec des /. https://deploy-preview-1037--nosgestesclimat.netlify.app/simulateur/bilan/transport/voiture/km.

Hmm j'ai rêvé ou à un moment on avait des urls du type `/simulateur/bilan/transport/dottedname-lié-par-des-tirets` ? En tout cas, des URLs de ce type feraient plus sens ; l'URL est sensée représenter l'architecture du site, sous dossier par sous dossier. En SEO il vaut mieux rester dans les clous, y'a déjà beaucoup de paramètres à prendre en compte ça vaut pas le coup de se rajouter un handicap à mon sens 😕 du coup dans l'idéal je verrais bien :
* `/simulateur/bilan/[categorie]` pour chaque catégorie permettant d'accéder au "mini" simulateur correspondant
* `/simulateur/bilan/[categorie]/[dottename]` pour chaque question, qu'en penses-tu ?

Concernant les query params : je suis pas convaincu que ce soit une super bonne pratique, à mon sens on fera mieux de faire les choses "classiquement" avec une url par page sans query params quand on peut l'éviter - l'URL de la page de fin avec `details` en est une bonne utilisation par exemple.

> je ne suis pas convaincu par le piège de boutons précédent. Après avoir répondu à 10 questions, je devrais cliquer 10 fois sur précédent pour revenir à la page d'accueil, si je n'ai pas envie d'utiliser le menu. Et idem sans raccourci si je partais de gouvernement.fr et que je voulais revenir sur gouvernement.fr. Ça ne me semble pas être l'objectif du bouton back. Comme expliqué ici https://github.com/datagir/nosgestesclimat-site/issues/463. Mais c'est pas bloquant, on peut aussi tester comme ça, mais on ne captera pas les frustrations potentielles des utilisateurs.

Plutôt pour le mode du fonctionnement du bouton précédent du navigateur pour le coup ! On serait dans les clous de ce qui est la norme.

File: source/sites/publicodes/Simulateur.tsx (hunk 21 above)
Comment (bjlaa) [reply source/sites/publicodes/Simulateur.tsx:235]:
```suggestion
			Sentry.captureException(e)
```

File: source/components/conversation/conversationUtils.ts (hunk 8 above)
Comment (EmileRolley) [reply source/components/conversation/conversationUtils.ts:69]: Pour signifier que la fonction utilisée sera `dispatch` ou (exclusif) `navigate` or avec un argument de type `{ dispatch?: ..., navigate?: ... }` les deux pourraient être définis ou non définis. Ici la contrainte assure que : 

1. pas plus d'une fonction sera renseignée,
2. au moins une sera renseignée,

(soit qu'une unique fonction sera disponible). De plus, cela permet de différencier les deux types de fonctions, ce qui ne serait pas possible si on avait qu'une seule fonction `toUse: Dispatch<AnyAction> | NavigateFunction`. Enfin du moins je n'ai pas trouver de solution pour ce problème.. On touche les limitation de TypeScript ici.

File: source/components/conversation/conversationUtils.ts (hunk 9 above)
Comment (EmileRolley) [reply source/components/conversation/conversationUtils.ts:89]: Hmm, je n'ai pas touché à la logique de cette fonction. Ta proposition est moins contraignante, en effet, une question de type `transport . empreinte divers` sera gardée dans pour `focusedCategory == 'divers'` alors que la question fait partie de la catégorie `transport` : l'implicite ici est que le premier mot de la règle définie sa catégorie.

File: source/components/publicodesUtils.tsx (hunk 10 above)
Comment (EmileRolley) [reply source/components/publicodesUtils.tsx:122]: C'est la norme dans NGC visiblement. C'est pour correspondre aux attributs publicodes.

File: source/sites/publicodes/chart/InlineCategoryChart.tsx (hunk 16 above)
Comment (EmileRolley) [reply source/sites/publicodes/chart/InlineCategoryChart.tsx:30]: Hmm personnellement je trouve ça très compréhensible, après je suis peut-être biaisé par le fait d'avoir l'habitude d'utiliser des langages où l’utilisation d'opérateurs _pipe_ est très fréquente.

File: source/sites/publicodes/chart/InlineCategoryChart.tsx (hunk 17 above)
Comment (EmileRolley) [reply source/sites/publicodes/chart/InlineCategoryChart.tsx:47]: > Pourquoi faut-il remaper les abbréviations à chaque modification de la situation ?

Ca ne semble pas nécessaire, en effet. Je même pas que ce soit nécessaire de le faire à chaque modification de règles car actuellement il n'y a pas de cas dans lequel une mise à jours des règles introduit une nouvelle catégorie non ? Il y a-t-il une raison historique peut-être @laem ?

File: source/sites/publicodes/Simulateur.tsx (hunk 18 above)
Comment (EmileRolley) [reply source/sites/publicodes/Simulateur.tsx:140]: Ce sont les descriptions des questions du modèle. Quel est le risque de dépasser ?

File: source/sites/publicodes/Simulateur.tsx [hunk 22]
```diff
@@ -0,0 +1,339 @@
...
+					</motion.div>
+				)}
+				{!isMainSimulation && (
+					<h1>
+						{evaluation.rawNode.title || (
+							<FullName dottedName={evaluation.rawNode.dottedName} />
+						)}
+					</h1>
+				)}
+				{!displayTutorial ? (
```
[the remark this answers is not in the document]
Comment (EmileRolley) [reply source/sites/publicodes/Simulateur.tsx:164]: Hmm je trouve ça déjà plus propre d'utiliser l'opérateur ternaire plutôt que la conjonction booléenne sur des variables qui ne le sont pas..

File: source/sites/publicodes/Simulateur.tsx (hunk 21 above)
Comment (EmileRolley) [reply source/sites/publicodes/Simulateur.tsx:235]: Pourquoi, on a pas envie de report une erreur à chaque fois que quelqu'un se trompe d'URL non ?

File: source/components/conversation/conversationUtils.ts (hunk 8 above)
Comment (bjlaa) [reply source/components/conversation/conversationUtils.ts:69]: Okay je vois ! Alors déjà d'instinct j'aurais plutôt choisi un nom descriptif du genre `handleSubmit`, généralement c'est une bonne pratique de nommer les méthodes et fonctions en mettant en évidence leur action, avec un nom commençant donc par un verbe (`getMachin`, `handleTruc`, `setBidule`). Pour en revenir au typage TS, je ne pense pas qu'il faille que les choix quant à la logique programmatique découle du bon typage, surtout quand ça apporte de la complexité au code 😕 je serais plutôt parti sur `nomDeLaMethode: Dispatch<AnyAction> | NavigateFunction`, qu'est-ce qui te parait bloquant ici ?

File: source/components/conversation/conversationUtils.ts (hunk 9 above)
Comment (bjlaa) [reply source/components/conversation/conversationUtils.ts:89]: D'acc !

File: source/components/publicodesUtils.tsx (hunk 10 above)
Comment (bjlaa) [reply source/components/publicodesUtils.tsx:122]: Bien noté !

File: source/sites/publicodes/Simulateur.tsx (hunk 18 above)
Comment (bjlaa) [reply source/sites/publicodes/Simulateur.tsx:140]: Pas de conséquences dramatiques, les moteurs de recherches coupent les descriptions plus longues que 320 caractères, simplement c'est considéré comme une bonne pratique de rester dans ces clous pour que l'utilisateur puisse lire l'entièreté de la description.

File: source/sites/publicodes/Simulateur.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (bjlaa) [reply source/sites/publicodes/Simulateur.tsx:164]:
```suggestion
				{ 
					!displayTutorial && !displayScoreExplanation && (
						<Simulation
							conversationProps={{
								orderByCategories: categories,
								customEnd: isMainSimulation ? (
									<MainSimulationEnding {...{ rules, engine }} />
								) : categoryRule.description ? (
									<Markdown
										children={categoryRule.description}
										noRouter={false}
									/>
								) : (
									<EndingCongratulations />
								),
							}}
							explanations={<InlineCategoryChart givenEngine={undefined} />}
						/>
					)
				}
				{displayTutorial && (
					<TutorialRedirection
						selectedRuleDottedName={selectedRuleDottedName}
						selectedRuleURL={selectedRuleURL}
					/>
				)}
```
J'ai l'impression que c'est un peu plus clair en découpant les conditions de cette manière, qu'en penses-tu ?

File: source/sites/publicodes/Simulateur.tsx (hunk 21 above)
Comment (bjlaa) [reply source/sites/publicodes/Simulateur.tsx:234]: Hmm à discuter ! Je me demande dans quels cas de figure on pourrait avoir une URL erronée 🤔 soit si on tape directement dans la barre de recherche, soit si on a un lien erroné. Effectivement le premier cas n'est pas intéressant à suivre, le deuxième peut-être plus. Bon ce cas-là n'est peut-être pas le plus pertinent mais en tout cas aux endroits stratégiques je pense qu'il faut pas hésiter à l'ajouter !

File: source/components/conversation/conversationUtils.ts (hunk 8 above)
Comment (EmileRolley) [reply source/components/conversation/conversationUtils.ts:69]: Je suis d'accord avec toi pour le nommage et `toUse` me parait clair avec le nom de la fonction non ? En revanche, pour moi toute l'utilité des types et de restreindre le nombre de comportements possible d'une fonction à ceux que l'on souhaite. Et comme je l'ai expliqué le fait d'utiliser une fonction au lieu d'un objet ne permet pas de conditionner sur le type de la fonction..

File: source/sites/publicodes/chart/InlineCategoryChart.tsx (hunk 16 above)
Comment (laem) [reply source/sites/publicodes/chart/InlineCategoryChart.tsx:30]: Le seul point qui me semble problématique c'est quand on veut logguer ou analyser la variable avant le map. Pas d'avis du coup, j'utilise les eux. 

Ici j'aurais même fait une arrow function pour éviter de créer un bloc.

File: source/sites/publicodes/chart/InlineCategoryChart.tsx (hunk 17 above)
Comment (laem) [reply source/sites/publicodes/chart/InlineCategoryChart.tsx:47]: Bonne question ! En effet, ça n'a pas d'intérêt. Cela dit c'est une opération anodine en terms de temps de calcul.

Comment (laem) [reply source/sites/publicodes/chart/InlineCategoryChart.tsx:47]: ça pourrait être mis dans extractCategories je pense, mais possible que ce soit le seul endroit où on en ait l'usahe.

**Comment by bjlaa** [comment]
Désolé mais la PR est vraiment colossale, c'est compliqué à relire ; est-ce que tu as besoin qu'on repasse ensemble sur certains points sur lesquels tu aurais des doutes ?

Sinon, j'ai fait quelques tests c'est trop cool ! Juste un détail au niveau du formatage de la valeur de paramètre pour la question : j'ai par exemple `?question=alimentation.de-saison.consommation` j'ai l'impression que les underscore sont plus courants dans des urls que des points, tu crois pas ? 🤔 je me demande aussi si ce serait pas plus lisible côté Matomo

File: source/components/publicodesUtils.tsx (hunk 11 above)
Comment (EmileRolley) [reply source/components/publicodesUtils.tsx:187]: Ca permet d'avoir que une string dans parent pas potentiellement un boolean

**Comment by EmileRolley** [comment]
> Juste un détail au niveau du formatage de la valeur de paramètre `question` : j'ai par exemple `?question=alimentation.de-saison.consommation` j'ai l'impression que les underscore sont plus courants dans des urls que des points, tu crois pas ? thinking je me demande aussi si ce serait pas plus lisible côté Matomo

Selon la [RFC 2396](https://www.ietf.org/rfc/rfc2396.txt), les deux symboles `.` et `_` sont des signes de ponctuations non-réservé est donc valide dans une URL. La question après est peut-être de savoir si on veut unifier le format avec celui utilisé pour la sauvegarde en BD, car il me semble que les points n'étaient pas acceptés ?

File: source/components/conversation/conversationUtils.ts (hunk 8 above)
Comment (EmileRolley) [reply source/components/conversation/conversationUtils.ts:74]: Je pense que l'on pourra revenir sur ce détail si on se rend compte que ça pose problème par la suite, mais pour l'instant je pars du principe que la signature de la fonction est suffisamment explicite pour que ne ce soit pas un réel problème.

File: source/components/publicodesUtils.tsx (hunk 12 above)
Comment (EmileRolley) [reply source/components/publicodesUtils.tsx:282]: A investiguer dans une prochaine PR de refactoring

File: source/sites/publicodes/Simulateur.tsx (hunk 14 above)
Comment (EmileRolley) [reply source/sites/publicodes/Simulateur.tsx]: J'ai rajouté un commentaire.

**Comment by laem** [comment]
De mon côté, je trouve que le _ pour remplacer le ` . ` serait trompeur car le _ est en général utilisé pour remplacer les espaces quand on a besoin d'encoder. 

@EmileRolley tu en es où dans cette PR ? Possibilité de mettre en ligne aujourd'hui ?

**Pull request merged**